Fix #2936: Limit the length of goal description in explorations.

shaz13 opened the pull request:
#2936
This PR is a suggested fix in the issue above. The exploration description is limited to 100 charaters

seanlip commented:
Hi @shaz13, one question: let's say that an existing exploration in the production webapp has an objective that's currently > 100 chars. After your change is merged and the release is pushed, what would the creator experience be like?

(You can test by checking out develop and creating an exploration with a long objective -- then checking out your feature branch and going to your existing exploration without restarting the server.)

shaz13 commented:
Sure, Sir. 
**Expected behaviour**: This change must reflect to the newly created explorations. The previous explorations will have this PR's effect only when one edit's the Settings of that exploration.
Thanks @seanlip  for suggestions. Will test on my side as suggested.

shaz13 commented:
@seanlip  Experiment done as suggested. 
Before this PR's commit one can go above 100 characters on Goal description. 

![selection_005](https://cloud.githubusercontent.com/assets/24438869/24638129/bf164558-1903-11e7-93a7-b2bb0ad534b5.png)

After the PR's commit the Goal description is unchanged. 

![selection_007](https://cloud.githubusercontent.com/assets/24438869/24638148/e992c018-1903-11e7-8ab3-1157fc9852b5.png)

Seems, as expected behaviour but, one cannot save the above 100 char description after merging PR. Example, in the second case though the description remains the same. User cannot publish the exploration until he modifies it around 100 or below characters.

shaz13 commented:
A red mark around the input text box is hinting a user an error. I guess instead of a place holder I need to modify instructions below the text area. From 

> What does this exploration help people to do? Please provide a complete, descriptive sentence.

> What does this exploration help people to do? Please provide a complete, descriptive sentence in about 100 characters.

shaz13 commented:
@jaredsilver  Sure. That will be more convenient to the user as one can frame sentences with remaining word limit.
A quick confirmation, Is jquery code allowed at Oppia or it is strictly JavaScript?

jaredsilver commented:
@shaz13 If you can do it using Angular, that would probably be preferable.

shaz13 commented:
@jaredsilver Kindly check the pushed commit with respect to the expected behaviour. 
Please note that the present commit uses JavaScript, I am currently getting errors while implementing it in angular JS would need a little help to cross check the code.

shaz13 commented:
The Angular Js code: 

```
<div class="col-lg-10 col-md-10 col-sm-10">
       <input id="explorationObjective" type="text"  class="form-control protractor-test-exploration-objective-input"
             ng-model="explorationObjectiveService.displayed"
                       ng-blur="saveExplorationObjective()" placeholder="Learn how to ..." ng-maxlength="{{maxLength}}>
                <span class="help-block" style="font-size: smaller">
                  What does this exploration help people to do? <span ng-if="explorationObjectiveService.displayed.length > 15" id="counter"></span>
    <span>Characters left: {{maxLength - explorationObjectiveService.displayed.length}}</span>
                  <span ng-if="!explorationObjectiveService.displayed || explorationObjectiveService.displayed.length < 15">Please provide a complete, descriptive sentence.</span>
                </span>
              </div>
```
and at controller: 
```

.controller('myCtrl' /* other conrollers here */, function($scope /* other dependencies here */){
    ...
    $scope.maxLength = 100; // this is the max # of chars allowed in the goal decription 
    ...
});
```

seanlip commented:
I haven't had time to look into how to do this, but I just wanted to echo @jaredsilver's note that the code should use Angular constructs. We do not allow innerHTML due to potential security issues.

shaz13 commented:
@seanlip  I am near to Angular JS solution but, experiencing blank page and errors. Seems like error with getting **explorationObjectiveService.displayed.length** value

shaz13 commented:
It shows Error: length of undefined. for **explorationObjectiveService.displayed.length**. When I use it in $scope.counter = 100 - explorationObjectiveService.displayed.length;

shaz13 commented:
@jaredsilver 
Done. Used inline script along angular class. Kindly review it
![selection_014](https://cloud.githubusercontent.com/assets/24438869/24702803/2e5c617a-1a1e-11e7-9eb8-0db77d42bf03.png)

seanlip commented:
@shaz13 -- happy to review again once you've addressed @jaredsilver's comments, but just wanted to say that from a code perspective this looks much better! Thank you for refactoring.

shaz13 commented:
@jaredsilver I agree with points 2 and 3. 
Regarding point 1, to be able to show over limit of characters I need to remove `max-length` property, as current commit doesn't allow above 100 characters at all. 
And regarding point 2, for clarification
White spaces should also should make count in char length ? Did I get it right? If not please do explain. 
Point 3 is crystal clear. Will push a commit later today. 
Thanks

shaz13 commented:
@jaredsilver The commit consist the suggested changes for all the three points. 

**Commit Notes:** 

1. Removed max-length (as per point no. 1)
2. Added inline code: `((explorationObjectiveService.displayed.length < 15) ||(explorationObjectiveService.displayed.length > 100))`

To meet the expected behaviour as addressed; 
> If an exploration already has a goal description of length greater than 100, the 100 character limit does not display. It would be ideal if the character limit also showed if the number of characters is greater than the limit. This way, people will know to shorten what they've written.

**Less than 15 chars**

![screenshot from 2017-04-06 19-44-08](https://cloud.githubusercontent.com/assets/24438869/24759918/9dd4c100-1b04-11e7-8559-6518ad4ff79e.png)

**Greater than 100 chars**

![screenshot from 2017-04-06 19-45-44](https://cloud.githubusercontent.com/assets/24438869/24759930/a93ce7d4-1b04-11e7-914e-ba5ab6519234.png)


3. Added extra angular inline `data-ng-trim="false"` , now the spaces are counted dynamically. Thanks for pointing it out :+1: 

> The number does not increment with whitespace characters on either end of the string, likely because we're trimming leading and trailing whitespace somewhere. It would probably be good to count at least trailing whitespace in the character count so creators can see the number increment as expected when they type a space.

For third point refactored the placeholder from **_Learn how to.._.** to **_What does this exploration help people to do_** following the convection of Title's input and span error as shown below

**Place holder and span error text changes** 

![screenshot from 2017-04-06 18-49-47](https://cloud.githubusercontent.com/assets/24438869/24759997/e44c600c-1b04-11e7-9577-ab905324192b.png)

Any further changes are welcome @seanlip 
Thanks

shaz13 commented:
@jaredsilver  Agreed cent percent. Will push the commits soon.

shaz13 commented:
@jaredsilver Done. Please review the PR :smile_cat:

shaz13 commented:
@jaredsilver  Done. Kidnly review the PR

shaz13 commented:
@jaredsilver Done as suggested. And spaces was not counted dynamically and the code had repeated logic. Minified all of those and pushed a commit on your suggestions. Kindly review it.

jaredsilver commented:
Thanks @shaz13! I apologize for the delay on this, but I'll have to get around to reviewing this tomorrow.

shaz13 commented:
@jaredsilver NP. I can wait. Have a good day. :smile:

seanlip commented on core/templates/dev/head/pages/exploration_editor/settings_tab/settings_tab.html:
```diff
@@ -8,22 +8,27 @@
+                <span class="help-block" style="font-size: smaller"> 
+                    <em>In a complete sentence, tell people what they'll learn from this exploration.</em>
+                    <span ng-if="explorationObjectiveService.displayed.length > 0">
+                    <em><[ "Characters used: " +  (explorationObjectiveService.displayed.length) + "/100" ]></em>
```
Indent by 2.

seanlip replied:
Also, move the "Characters used" etc. out of the angular brackets:

```
Characters used: <[...]>/100
```

shaz13 replied:
Indent by 2, is it for the second **em** tag? Kindly confirm

seanlip replied:
I don't understand your question. But this em tag is contained within the span tag on the previous line, if I understand correctly. Since it is a child of span, and not a sibling, it should be indented by 2.

shaz13 replied:
Got it. Thanks

seanlip commented on core/templates/dev/head/pages/exploration_editor/settings_tab/settings_tab.html:
```diff
@@ -8,22 +8,27 @@
               <label for="explorationTitle" class="col-lg-2 col-md-2 col-sm-2">Title</label>
               <div class="col-lg-10 col-md-10 col-sm-10">
-                <input id="explorationTitle" type="text" class="form-control protractor-test-exploration-title-input" ng-model="explorationTitleService.displayed" ng-blur="saveExplorationTitle()" placeholder="Choose a title for your exploration." focus-on="<[::EXPLORATION_TITLE_INPUT_FOCUS_LABEL]>" maxlength="40">
+                <input id="explorationTitle" type="text" class="form-control protractor-test-exploration-title-input" ng-model="explorationTitleService.displayed" ng-blur="saveExplorationTitle()" placeholder="Choose a title for your exploration." focus-on="<[::EXPLORATION_TITLE_INPUT_FOCUS_LABEL]>" maxlength="40" data-ng-trim="false">
```
What is the effect of data-ng-trim?

shaz13 replied:
It allows whitespace before/after the string to count dynamically. Default value is true which results the counting of space only after a character. 
Example: V     H (Here the value of characters are updated after H is typed, i.e 5 to 10)
With data-ng-trim, V    H  (The values are incremented space by space )

seanlip replied:
I can't find it in the docs: https://docs.angularjs.org/api/ng/input/input%5Btext%5D . Can you provide a reference?

Also, why not just ng-trim?

shaz13 replied:
Sure,
https://stackoverflow.com/questions/16589853/ng-app-vs-data-ng-app-what-is-the-difference.
It makes page more HTML friendly.

seanlip replied:
OK, then just use ng-trim please. Otherwise this is the only part in the entire application that is more html friendly :)

If we want to do this sort of change, it needs to happen everywhere, for codebase consistency.

shaz13 replied:
Sure.

shaz13 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/settings_tab/settings_tab.html:
```diff
@@ -8,22 +8,27 @@
                 <span class="help-block" style="font-size: smaller">
                   <em>Please use at most 40 characters, so that this fits in the summary card.</em>
+                  <span ng-if="explorationTitleService.displayed.length > 0">
+                  <em><[ "Characters used: " +  (explorationTitleService.displayed.length) + "/40" ]></em>
```
See comment below.

seanlip commented on core/templates/dev/head/pages/exploration_editor/settings_tab/settings_tab.html:
```diff
@@ -8,22 +8,27 @@
-                  <span ng-if="!explorationObjectiveService.displayed || explorationObjectiveService.displayed.length < 15">Please provide a complete, descriptive sentence.</span>
+                       ng-blur="saveExplorationObjective()" placeholder="Learn how to..." data-ng-trim="false">
+                <span class="help-block" style="font-size: smaller"> 
+                    <em>In a complete sentence, tell people what they'll learn from this exploration.</em>
```
Indentation is wrong. Should be by 2 from the previous line.

shaz13 commented:
Sure,
https://stackoverflow.com/questions/16589853/ng-app-vs-data-ng-app-what-is-the-difference.
It makes page more HTML friendly.

On 11 April 2017 at 03:53, Sean Lip <notifications@github.com> wrote:

> *@seanlip* commented on this pull request.
> ------------------------------
>
> In core/templates/dev/head/pages/exploration_editor/settings_
> tab/settings_tab.html
> <https://github.com/oppia/oppia/pull/3274#discussion_r110777117>:
>
> > @@ -8,22 +8,27 @@
>              <div class="form-group" ng-class="{'has-error': !explorationTitleService.displayed}">
>                <label for="explorationTitle" class="col-lg-2 col-md-2 col-sm-2">Title</label>
>                <div class="col-lg-10 col-md-10 col-sm-10">
> -                <input id="explorationTitle" type="text" class="form-control protractor-test-exploration-title-input" ng-model="explorationTitleService.displayed" ng-blur="saveExplorationTitle()" placeholder="Choose a title for your exploration." focus-on="<[::EXPLORATION_TITLE_INPUT_FOCUS_LABEL]>" maxlength="40">
> +                <input id="explorationTitle" type="text" class="form-control protractor-test-exploration-title-input" ng-model="explorationTitleService.displayed" ng-blur="saveExplorationTitle()" placeholder="Choose a title for your exploration." focus-on="<[::EXPLORATION_TITLE_INPUT_FOCUS_LABEL]>" maxlength="40" data-ng-trim="false">
>
> I can't find it in the docs: https://docs.angularjs.org/
> api/ng/input/input%5Btext%5D . Can you provide a reference?
>
> Also, why not just ng-trim?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/3274#discussion_r110777117>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXToVRnx3p_o-3-r_hL7jcONh7zQZFzKks5ruqvHgaJpZM4MyRhx>
> .
>

shaz13 commented:
@seanlip  Done as suggested kindly review again.

shaz13 commented:
@seanlip Do this PR needs any other changes? Can we label **LGTM** batch? or Other reviews are required?

seanlip commented:
It's good, thanks! Merging it.
